minvws/nl-kat-coordination
# Fix Garbage collection and disappearing ports issue

### Changes

This, ideally, fixes the garbage collection issue with `nmap` and `nmap-udp`.

It also adds a rather involved migration script that should still be thoroughly tested somehow. There we take all origins that should be updated and add data from bytes to the new `source_method` field. (The name for that field is very much to be discussed, by the way.) Note that it has been added to the `boefjes` module because it has connections to both Bytes and Octopoes already and I guess I'm more comfortable there. I wouldn't mind if there's a solid argument for using Rocky instead. 

### Issue link

Closes #2875

### QA notes

Follow the instructions in #2875 and hopefully the issue is now gone.

However, there is a much more intricate part to QA here: the migration script. We need to update the existing origins using Octopoes and Bytes at the same time. This means: getting a realistic working setup, building, and triggering the migration using Docker:

```
git checkout fix/disappearing-ports
git pull
make kat
docker compose exec boefje python -m tools.upgrade_v1_16_0
```

Now the issue should not be present anymore.

The logs ideally show no exceptions and end with `total_failures=0`. If not, it can be rerun multiple times in case a random network error appeared. I say this because we could potentially hit the complete database and hence the script had a high chance of failing some API calls in production systems. If there are consistent exceptions however, there is probably a bug.

---

UPDATE:

Benchmark without new filter:

```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.001    0.001    7.903    7.903 /app/boefjes/tests/integration/test_bench.py:19(test_migration)
        1    0.000    0.000    4.364    4.364 /app/boefjes/tools/upgrade_v1_16_0.py:39(upgrade)
        1    0.003    0.003    3.953    3.953 /app/boefjes/tools/upgrade_v1_16_0.py:71(migrate_org)
       30    0.001    0.000    3.193    0.106 /app/boefjes/tools/upgrade_v1_16_0.py:133(update_origin)
        1    0.000    0.000    2.362    2.362 /app/boefjes/tests/conftest.py:1(<module>)
        2    0.000    0.000    1.362    0.681 /app/boefjes/tests/conftest.py:173(octopoes_api_connector)
        1    0.000    0.000    1.314    1.314 /app/boefjes/boefjes/app.py:1(<module>)
      121    0.000    0.000    0.879    0.007 /app/boefjes/boefjes/clients/bytes_client.py:20(wrapper)
        2    0.000    0.000    0.773    0.386 /app/boefjes/boefjes/clients/bytes_client.py:48(login)
        2    0.000    0.000    0.773    0.386 /app/boefjes/boefjes/clients/bytes_client.py:62(_get_authentication_headers)
```

Benchmark with new filter:
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.001    0.001    5.428    5.428 /app/boefjes/tests/integration/test_bench.py:19(test_migration)
        1    0.000    0.000    2.399    2.399 /app/boefjes/tests/conftest.py:1(<module>)
        1    0.000    0.000    1.427    1.427 /app/boefjes/tools/upgrade_v1_16_0.py:40(upgrade)
        2    0.000    0.000    1.410    0.705 /app/boefjes/tests/conftest.py:173(octopoes_api_connector)
        1    0.000    0.000    1.349    1.349 /app/boefjes/boefjes/app.py:1(<module>)
        1    0.000    0.000    1.012    1.012 /app/boefjes/tools/upgrade_v1_16_0.py:72(migrate_organisation)
        1    0.001    0.001    0.823    0.823 /app/boefjes/tools/upgrade_v1_16_0.py:147(collect_boefjes_per_normalizer)
        1    0.000    0.000    0.821    0.821 /app/boefjes/boefjes/dependencies/plugins.py:55(_get_all_without_enabled)
        1    0.000    0.000    0.806    0.806 /app/boefjes/boefjes/local_repository.py:28(get_all)
      100    0.000    0.000    0.802    0.008 /app/boefjes/boefjes/clients/bytes_client.py:20(wrapper)
```

Note that the `migrate_org` method first takes almost 4 seconds, and then this drops to 1 second, of which 800 ms is from `collect_boefjes_per_normalizer` which we only call once. Actually calling the new bulk endpoint takes only 28 ms for 30 origins. Still 30 origins per 200 ms for a 30% rate of to-be-updated origins (e.g. nmap) means we could do 150 origins per second in my setup.

---

### Code Checklist







---

## Checklist for code reviewers:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_code.md) into your comment.

---

## Checklist for QA:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_qa.md) into your comment.
